panda/src/navmeshgen/navMeshBuilder.cxx

```diff
@@ -0,0 +1,738 @@
+      vertex.set_row(a);
+      v = vertex.get_data3();
+      a = vertex_map[v];
+
+      int b = prim->get_vertex(s + 1);
+      vertex.set_row(b);
+      v = vertex.get_data3();
+      b = vertex_map[v];
+
+      int c = prim->get_vertex(s + 2);
+      vertex.set_row(c);
+      v = vertex.get_data3();
+      c = vertex_map[v];
+
+
+      LVector3 xvx = { float(a + 1), float(b + 1), float(c + 1) };
+      face_vector.push_back(xvx);
+      add_triangle(a, b, c, tcap);
+
+    }
+    else if (e - s > 3) {
+
+      for (int i = s + 2; i < e; ++i) {
+        int a = prim->get_vertex(s);
+        vertex.set_row(a);
+        v = vertex.get_data3();
+        a = vertex_map[v];
+
+        int b = prim->get_vertex(i - 1);
+        vertex.set_row(b);
+        v = vertex.get_data3();
+        b = vertex_map[v];
+
+        int c = prim->get_vertex(i);
+        vertex.set_row(c);
+        v = vertex.get_data3();
+        c = vertex_map[v];
+
+        LVector3 xvx = { float(a + 1), float(b + 1), float(c + 1) };
+        face_vector.push_back(xvx);
+        add_triangle(a, b, c, tcap);
+
+      }
+    }
+    else continue;
+
+  }
+  return;
+}
+
+void NavMeshBuilder::process_vertex_data(const GeomVertexData *vdata, int &vcap) {
+  GeomVertexReader vertex(vdata, "vertex");
+  float x, y, z;
+
+  while (!vertex.is_at_end()) {
+
+    LVector3 v = vertex.get_data3();
+    x = v[0];
+    y = v[1];
+    z = v[2];
+    if (vertex_map.find(v) == vertex_map.end()) {
+      add_vertex(x, z, -y, vcap); //if input model is originally z-up
+      //add_vertex(x, y, z, vcap); //if input model is originally y-up
+
+      //vertex_map[v] = index_temp++;
+      LVector3 xvx = { v[0],v[2],-v[1] };
+      vertex_map[v] = index_temp++;
+      vertex_vector.push_back(v);
+    }
+
+  }
+  return;
+}
+
+
+
+void NavMeshBuilder::process_geom(CPT(Geom) geom, int& vcap, int& tcap) {
+
+  CPT(GeomVertexData) vdata = geom->get_vertex_data();
+
+  process_vertex_data(vdata, vcap);
+
+  for (size_t i = 0; i < geom->get_num_primitives(); ++i) {
+    CPT(GeomPrimitive) prim = geom->get_primitive(i);
+    process_primitive(prim, vdata, tcap);
+  }
+  return;
+}
+
+void NavMeshBuilder::process_geom_node(GeomNode *geomnode, int &vcap, int &tcap) {
+
+
+  if (geomnode->get_num_geoms() > 1) {
+    std::cout << "Cannot proceed: Make sure number of geoms = 1\n";
+    return;
+  }
+  for (size_t j = 0; j < geomnode->get_num_geoms(); ++j) {
+    CPT(Geom) geom = geomnode->get_geom(j);
+    process_geom(geom, vcap, tcap);
+  }
+  return;
+}
+
+
+
+
+
+void NavMeshBuilder::cleanup()
+{
+  delete[] _triareas;
+  _triareas = 0;
+  rcFreeHeightField(_solid);
+  _solid = 0;
+  rcFreeCompactHeightfield(_chf);
+  _chf = 0;
+  rcFreeContourSet(_cset);
+  _cset = 0;
+  rcFreePolyMesh(_pmesh);
+  _pmesh = 0;
+  rcFreePolyMeshDetail(_dmesh);
+  _dmesh = 0;
+  
+}
+
+void NavMeshBuilder::set_partition_type(std::string p) {
+  if (p == "watershed" || p == "Watershed" || p == "WATERSHED") {
```

Comment:
`string_utils.h` has a `downcase()` to convert a string to all lowercase, which would make this check nicer and more robust:

```cpp
if (downcase(p) == "watershed") {
...
}```

Reply:
done